senator/game.py

```diff
@@ -78,45 +97,93 @@ def nash_conv(self) -> float:
-        child_utility_matrix = get_utility_matrix(current_utility, votes)  # TODO: see what it takes to get the utility matrix
-        node.children[tuple(joint_action)] = Node(parent=node, utility_matrix=child_utility_matrix)
-
-    def solve_node_via_rm_plus(self, node: Node, iterations: int = 1000) -> None:
+    def add_child(
+        self,
+        node: Node,
+        joint_action: np.ndarray,
+        current_utility: np.ndarray,
+        votes: np.ndarray,
+    ) -> None:
+        child_utility_matrix = get_utility_matrix(
+            current_utility,
+            votes,
+            self.is_owner,
+            self.owner_loss,
+            self.other_loss,
+            self.owner_trashold,
+        )  # TODO: see what it takes to get the utility matrix
+        node.children[tuple(joint_action)] = Node(
+            parent=node, utility_matrix=child_utility_matrix
+        )
+
+    def solve_node_via_rm_plus(self, node: Node, iterations: int = 2000) -> None:
         for _ in range(iterations):
             node.rm_plus_step()
 
         if (conv := node.nash_conv()) >= 0.001:
             print(f"Nash Convexity is {conv}, something is wrong...")
 
-    def run_greedy_search(self, num_expansions: int = 10):
-        """
-        Run a greedy search algorithm, periodically expanding the tree. Each expansion is done by first solving the
-        node using the RM+ algorithm, sampling a joing action, and adding the child to the tree.
-        TODO: decide how to do the expanding. Baseline is to just do a path of length 5.
-        """
+    def solve_node_via_rm(self, node: Node, iterations: int = 2000) -> None:
+        for _ in range(iterations):
+            node.rm_step()
+
+        if (conv := node.nash_conv()) >= 0.001:
+            print(f"Nash Convexity is {conv}, something is wrong...")
+
+    def run_greedy_search(self, num_steps: int = 7) -> tuple:
+
+        actions_in_steps = []
+        strategies_in_steps = []
+
         current_node = self.root
         utility = self.initial_utility
-        for _ in range(num_expansions):
-            self.solve_node_via_rm_plus(current_node)
-            joint_action = current_node.sample_joint_action()
-            utility = get_utility(utility, joint_action, self.votes)
-            self.add_child(current_node, joint_action, utility, self.votes) #TODO: check if I want votes as parameter of Game
-            current_node = current_node.children[tuple(joint_action)]
- 
 
+        # print(f"Initial utility\nu: {utility.sum(axis=1)}")
 
+        for i in range(num_steps):
+            self.solve_node_via_rm(current_node)
+            strategies_in_steps.append(current_node.current_strategy)
+            joint_action = current_node.sample_joint_action()
+
+            actions_in_steps.append(joint_action)
+
+            utility = get_utility(
+                utility,
+                joint_action,
```

Comment:
Tady mi to hlásí typovací chybu, protože `current_node.sample_joint_action()` vrací `np.array`, ale `get_utility` čeká tohle jako `list[int]`.

Reply:
Přehodím na np.array